src/nexmark/queries/q12.rs

```diff
@@ -0,0 +1,150 @@
+use super::{process_time, NexmarkStream};
+use crate::{nexmark::model::Event, operator::FilterMap, Circuit, OrdZSet, Stream};
+
+///
+/// Query 12: Processing Time Windows (Not in original suite)
+///
+/// How many bids does a user make within a fixed processing time limit?
+/// Illustrates working in processing time window.
+///
+/// Group bids by the same user into processing time windows of 10 seconds.
+/// Emit the count of bids per window.
+///
+/// ```sql
+/// CREATE TABLE discard_sink (
+///   bidder BIGINT,
+///   bid_count BIGINT,
+///   starttime TIMESTAMP(3),
+///   endtime TIMESTAMP(3)
+/// ) WITH (
+///   'connector' = 'blackhole'
+/// );
+///
+/// INSERT INTO discard_sink
+/// SELECT
+///     B.bidder,
+///     count(*) as bid_count,
+///     TUMBLE_START(B.p_time, INTERVAL '10' SECOND) as starttime,
+///     TUMBLE_END(B.p_time, INTERVAL '10' SECOND) as endtime
+/// FROM (SELECT *, PROCTIME() as p_time FROM bid) B
+/// GROUP BY B.bidder, TUMBLE(B.p_time, INTERVAL '10' SECOND);
+/// ```
+
+type Q12Stream = Stream<Circuit<()>, OrdZSet<(u64, u64, u64, u64), isize>>;
+const TUMBLE_SECONDS: u64 = 10;
+
+fn window_for_process_time(ptime: u64) -> (u64, u64) {
+    let window_lower = ptime - (ptime % (TUMBLE_SECONDS * 1000));
+    (window_lower, window_lower + TUMBLE_SECONDS * 1000)
+}
+
+// This function enables us to test the q12 functionality without using the
+// actual process time, while the actual q12 function below uses the real
+// process time.
+// TODO: I originally planned to pass a FnMut closure for process_time that
+// just emits a new u64 each time it is called, but can't do this as the
+// closure of `flat_map_index` is Fn not FnMut, and would need to capture the
+// process_time closure. So right now, it's quite ugly: to avoid an `FnMut`,
+// I'm instead passing an optional vector of times with the assumption that
+// those times are indexed by the bid.auction.
+// There must be a better way without resorting to interior mutability? Anyway,
+// it works for the tests and is only used in the tests.
+fn q12_for_process_time(input: NexmarkStream, process_times: Option<Vec<u64>>) -> Q12Stream {
+    let bids_by_bidder_window = input.flat_map_index(move |event| match event {
+        // TODO: Can I call process_time() just once per batch, rather than for every Bid? How?
```

Comment:
Since the performance is looking good, I wouldn't bother.

Reply:
I see even better performance when increasing the number of generators to, so this is one of those queries where we struggle to produce inputs fast enough.  Nice!

Unrelated: most queries run faster when I set `input-batch-size` to 40K.  Shall we make that that a default or is this cheating somehow?  With 40K batch size, 6 workers, and 8 generators I get:

```
cargo bench --bench nexmark --features with-nexmark -- --first-event-rate=10000000 --max-events=100000000 --cpu-cores 6 --query q12 --num-event-generators 8 --source-buffer-size 10000 --input-batch-size 40000
    Finished bench [optimized + debuginfo] target(s) in 0.06s
     Running benches/nexmark/main.rs (target/release/deps/nexmark-db7e78eeb5fe8ae0)
Starting q12 bench of 100000000 events...
100,000,000 / 100,000,000 [==========================================================================================================================================] 100 % 6672068.9755/s 0s
┌───────┬─────────────┬───────┬─────────┬─────────────────┬──────────────────┬───────────────┬───────────────┬─────────────┐
│ Query │ #Events     │ Cores │ Elapsed │ Cores * Elapsed │ Throughput/Cores │ Total Usr CPU │ Total Sys CPU │ Max RSS(Kb) │
├───────┼─────────────┼───────┼─────────┼─────────────────┼──────────────────┼───────────────┼───────────────┼─────────────┤
│ q12   │ 100,000,000 │ 6     │ 14.989s │ 89.932s         │ 1111.955 K/s     │ 170.236s      │ 4.562s        │ 414,800     │
└───────┴─────────────┴───────┴─────────┴─────────────────┴──────────────────┴───────────────┴───────────────┴─────────────┘

```

Reply:
In fact, with q0 (which doesn't even do string copying - literally a noop) with the same options I get 29s (or 32s with `--input-batch-size 1000`). I assume you get way under 15s with q0? The fact that you're getting only 15s with q12 also means that there's only one or two process windows during the benchmark (perhaps not an issue, and we can easily change the size of the window for q12 if we want).